### DISCUSSION TITLE
Built-in controllers dispose

### DISCUSSION THREAD

**Discussion by busslina** [GregoryConrad/rearch-dart#178] [General]
Is there a reason on built-in controllers (`TextEditingController`, `FocusNode` and `FocusScopeNode`) to not being auto-disposed?

For this purpose I also envision that would be helpful to have a built-in side effect that execute once (like callonce), return the value and also execute on disposing. Similar to `use.effect` but this last one is void.

**Comment by GregoryConrad** [comment]
They should be getting disposed; see https://github.com/GregoryConrad/rearch-dart/blob/main/packages/flutter_rearch/lib/src/side_effects/text_editing_controller.dart

After looking at the current FocusNode effect source they aren't getting disposed. That's a bug.

> For this purpose I also envision that would be helpful to have a built-in side effect that execute once (like callonce), return the value and also execute on disposing. Similar to use.effect but this last one is void.

Yea, like a:

```dart
use.disposable(FocusNode.new, (node) => node.dispose());
```

Assuming Dart's type inference doesn't screw that up, that seems like a solid idea to me (and could be used to fix the FocusNode effects

**Reply by busslina** [reply]
> `use.disposable(FocusNode.new, (node) => node.dispose());`

I like it.

**Comment by satvikpendem** [comment]
Would this be similar to Riverpod's `autoDispose`?

**Reply by busslina** [reply]
Dont think so. Riverpod auto dispose refers to a provider being disposed when it has no active listeners. Here is not about a capsule dispose, it's about a controller being disposed on capsule dispose.
